src/runtime/components/server/component_base.cpp

```diff
@@ -0,0 +1,121 @@
+#include <hpx/runtime/naming/name.hpp>
+
+namespace hpx { namespace detail
+{
+    HPX_API_EXPORT naming::gid_type get_next_id(std::size_t count = 1);
+}}
+
+namespace hpx { namespace components { namespace detail
+{
+    component_base::~component_base()
+    {
+        if (gid_)
+        {
+            error_code ec;
+            agas::unbind(launch::sync, gid_, 1, ec);
+        }
+    }
+
+    naming::gid_type component_base::get_base_gid(
+        naming::gid_type const& assign_gid) const
+    {
+        if (!gid_)
+        {
+            naming::address addr(get_current_address());
+            if (!assign_gid)
+            {
+                gid_ = hpx::detail::get_next_id();
+                if (!applier::bind_gid_local(gid_, addr))
+                {
+                    std::ostringstream strm;
+                    strm << "failed to bind id " << gid_
+                            << "to locality: " << hpx::get_locality();
+
+                    gid_ = naming::invalid_gid;    // invalidate GID
+
+                    HPX_THROW_EXCEPTION(duplicate_component_address,
+                        "component_base<Component>::get_base_gid",
+                        strm.str());
+                }
+            }
+            else
+            {
+                applier::applier& appl = hpx::applier::get_applier();
+                gid_ = assign_gid;
+                naming::detail::strip_credits_from_gid(gid_);
+
+                if (!agas::bind(
+                        launch::sync, gid_, addr, appl.get_locality_id()))
+                {
+                    std::ostringstream strm;
+                    strm << "failed to rebind id " << gid_
+                            << "to locality: " << hpx::get_locality();
+
+                    gid_ = naming::invalid_gid;    // invalidate GID
+
+                    HPX_THROW_EXCEPTION(duplicate_component_address,
+                        "component_base<Component>::get_base_gid",
+                        strm.str());
+                }
+            }
+        }
+
+        std::unique_lock<naming::gid_type::mutex_type> l(gid_.get_mutex());
+
+        if (!naming::detail::has_credits(gid_))
+        {
+            naming::gid_type gid = gid_;
+            return gid;
+        }
+
+        // on first invocation take all credits to avoid a self reference
+        naming::gid_type gid = gid_;
+
+        naming::detail::strip_credits_from_gid(
+            const_cast<naming::gid_type&>(gid_));
+
+        HPX_ASSERT(naming::detail::has_credits(gid));
+
+        // We have to assume this credit was split as otherwise the gid
+        // returned at this point will control the lifetime of the
+        // component.
+        naming::detail::set_credit_split_mask_for_gid(gid);
+        return gid;
+    }
+
+    naming::address component_base::get_current_address() const
+    {
+        return naming::address(get_locality(), type_,
+            std::uint64_t(static_cast<component_base const*>(this)));
```

Comment:
the static cast here is not necessary, is it?

Reply:
It is needed as `this` might be different from the base address of the whole component.

Reply:
But this code is wrong for other reasons, will fix.